### PR TITLE
[express-serve-static-core] add RouteParameters type for IRouterMatcher that uses template literal types

### PR DESCRIPTION
**Pull request by DetachHead**
image

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:
If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html

### CODE REVIEW

**Comment by danger-public** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## express ([<kbd>unpkg</kbd>](https://unpkg.com/browse/express@latest/))
was missing the following properties:
1. Route
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 59dc475d2ed6032f892a777b8061c11c8d6fbeaa
</p>

**Comment by DetachHead** [comment]
the failed job appears to be due to an issue in the environment where the benchmark is run:

image

**Comment by sheetalkamat** [comment]
This is important package with change, needs approval from one of @borisyankov, @19majkel94, @kacepe, @micksatana, @samijaber, @JoseLion, @dwrss, @andoshin11, @CMUH, @puneetar, @dfrankland: to approve before this can be merged

**Comment by weswigham** [comment]
Yeah, at a surface level this looks pretty OK, but has the potential to slightly break people.

@borisyankov, @19majkel94, @kacepe, @micksatana, @samijaber, @JoseLion, @dwrss, @andoshin11, @CMUH, @puneetar, @dfrankland if one of you could show up and drop a review, that'd help a lot.

**Comment by orta** [comment]
Heh, 4th TypeScript engineer kicking this PR down the road. I agree with Wes' feedback, on the whole this looks reasonable and may introduce a break or two. If this was combined with bad perf, I'd be a bit more meh on the additional complexity for the feature but the stats seems fine.

**Comment by elibarzilay** [comment]
(Same.)

@borisyankov, @19majkel94, @kacepe, @micksatana, @samijaber, @JoseLion, @dwrss, @andoshin11, @CMUH, @puneetar, @dfrankland ???

**Review by JoseLion — APPROVED** [review]
I agree with @weswigham, @orta, and @elibarzilay! IMO, even if it may introduce a break or two, the increased type-safety in route params is worth the breaking changes. It also seems everything is very well tested, plus I cannot see a better way to take advantage of template literal types. 😀  

I say let's give it a go! 🙂

**Comment by andrewbranch** [comment]
I think this looks good and would love to see it merged. I would also love to see another approval, since it’s one of the most widely used set of typings in the world 😅

**Comment by sandersn** [comment]
Welllll, this package is *literally* the main reason that template literal types exist, so I think we should merge this PR eventually. I'm going to merge it in about 16 hours, so that I have all day to watch for complaints. @DetachHead is that a good time for you? If not, I'll be prepared to revert the merge so you can fix any bugs at your leisure.

**Comment by DetachHead** [comment]
I'll probably be asleep as that's around midnight for me. But I'll check first thing in the morning

File: types/express-serve-static-core/express-serve-static-core-tests.ts
Comment (robertvanhoesel) [inline types/express-serve-static-core/express-serve-static-core-tests.ts:109]: This seems like a decent fallback for non standard implementations. Would users still be able to provide custom parameters like in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/51262/files#diff-c5d8f61506d81e402ccb7f5f031730ce5ee044b86777af13924d7edfae4243f1R45

Comment (DetachHead) [reply types/express-serve-static-core/express-serve-static-core-tests.ts:109]: yeah, this works:
```ts
app.get<{ foo: string; bar: number }>('/' as string, req => {
    req.params.foo; // $ExpectType string
    req.params.bar; // $ExpectType number
    req.params.baz; // $ExpectError
});
```
i can add a test for this if you'd like

Comment (robertvanhoesel) [reply types/express-serve-static-core/express-serve-static-core-tests.ts:109]: This is great, good to have this workaround on hand for when it might break someones implementation (although it is hard to imagine how it would). This is a great PR!

File: types/express-serve-static-core/index.d.ts
Comment (danvk) [inline types/express-serve-static-core/index.d.ts:284]: Suggest using a more descriptive name than `T`. (See https://fettblog.eu/tidy-typescript-name-your-generics/)

File: types/express-serve-static-core/ts4.0/express-serve-static-core-tests.ts
Comment (danvk) [inline types/express-serve-static-core/ts4.0/express-serve-static-core-tests.ts:44]: nit: accommodate (spelling)

File: types/express-serve-static-core/ts4.0/express-serve-static-core-tests.ts
Comment (danvk) [inline types/express-serve-static-core/ts4.0/express-serve-static-core-tests.ts:161]: The Express docs have a few examples of tricky routes: https://expressjs.com/en/guide/routing.html#route-parameters

It would be a good idea to add them to the tests here to make sure they work as expected:
- `/flights/:from-:to`
- `/plantae/:genus.:species`
- `/user/:userId(\d+)`

While it's not explicitly mentioned on that page, if you click through to [`path-to-regexp`](https://www.npmjs.com/package/path-to-regexp), it mentions that parameters can be optional, i.e. `/path/:param?` should make the type of `req.params.param` be `string | undefined`.

This may be tested elsewhere, but the path can also be a regular expression, and that should continue to work as before (there are no regex literals… yet!).

**Review by danvk — COMMENTED** [review]
Overall very excited about this change! See comment about optional params in paths. I wouldn't want to block on that, but it would be worth seeing how hard it is to handle them correctly.

File: types/express-serve-static-core/ts4.0/express-serve-static-core-tests.ts
Comment (DetachHead) [reply types/express-serve-static-core/ts4.0/express-serve-static-core-tests.ts]: both this and the misspelling of "coersion" were already there in the original file https://github.com/DefinitelyTyped/DefinitelyTyped/pull/51262/files/59dc475d2ed6032f892a777b8061c11c8d6fbeaa#diff-459c5bd4ded180bff2de08f3ae44f73df64deea0d0ef3989e729876d43daa60c

i fixed it in both files

File: types/express-serve-static-core/ts4.0/express-serve-static-core-tests.ts
Comment (DetachHead) [reply types/express-serve-static-core/ts4.0/express-serve-static-core-tests.ts:161]: oh, wasn't aware of this. i'll add support for `-`, `.` and `?`, though the regex will probably just have to be ignored

Comment (danvk) [reply types/express-serve-static-core/ts4.0/express-serve-static-core-tests.ts:161]: How about just adding a test for `/path/:param?` so that we know what the behavior is after this PR? Even if we know it's incorrect, it will be a good place to capture the context from this discussion and what more correct behavior would be in the future.

Comment (DetachHead) [reply types/express-serve-static-core/ts4.0/express-serve-static-core-tests.ts:177]: done. though i'm keen to rewrite the `RouteParameters` type to support this as well as properly ignore regex wrapped in parentheses. i think it'll need to be more like a lexer. i'll get on that tomorrow

**Comment by sandersn** [comment]
@DetachHead you *probably* aren't still awake, but it looks like the PR is in a done state now, and the next chunk of work is a fairly large, separate thing. I'll merge this in a couple of hours in the off chance that you'll have a chance to object before then.

**Comment by orta** [comment]
I think the way that these changes were set up still allows TS versions < 4.0 to get the d.ts files for the template literals - https://github.com/DefinitelyTyped/DefinitelyTyped/issues/53397

**Comment by JensForstmann** [comment]
For anybody having trouble right now with optional parameters and getting an error at build time:

```ts
server.get('/download/:entity/:name?', async (req, res, next) => {
	const entity = req.params.entity;

	// used to work in the past prior this PR
	const nameOr = req.params.name; // <<-- error here: Property 'name' does not exist on type 'RouteParameters<"/:entity/preview/:name?">'. Did you mean 'name?'?

	// as suggested
	const nameOr = req.params['name?']; // build will work, but variable will always be undefined
	
	// ...
});
```

The workaround that worked for me is changing the route to:

```ts
// Workaround: add empty string: route + ''
server.get('/download/:entity/:name?' + '', async (req, res, next) => {
	// ...
});
```

So just by concatenating two strings as a route it seems that the work put in to this PR will get disabled.

The type of `req.params` is then `(property) Request<RouteParameters<string>, any, any, QueryString.ParsedQs, Record<string, any>>.params: RouteParameters<string>`

**Comment by DetachHead** [comment]
you can do this:
```ts
server.get<{entity: string, name?: string}>('/download/:entity/:name?', async (req, res, next) => {
	// ...
});
```
that way you still get the type safety and aren't adding any unnecessary runtime code.

sorry about that, i'll try to get a PR ready with a fix for this within the next few days

File: types/express-serve-static-core/ts4.0/express-serve-static-core-tests.ts
Comment (DetachHead) [reply types/express-serve-static-core/ts4.0/express-serve-static-core-tests.ts:168]: #53513

Comment (DetachHead) [reply types/express-serve-static-core/ts4.0/express-serve-static-core-tests.ts:168]: #53513

**Comment by DetachHead** [comment]
#53513

**Comment by angelsvirkov** [comment]
@DetachHead  is it really intended that projects using typescript versions below 4 should receive the type definitions from the `ts4.0` folder? 


Currently ,in the package.json of both [@types/express](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/express/package.json#L8) and [@types/express-serve-static-core](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/express-serve-static-core/package.json#L5)] the `typesVersions` property points to everything below `<=4` instead of `>=4` as normally done on type packages and defined in the [TS Docs](https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html#version-selection-with-typesversions)
```json
    "typesVersions": {
        "<=4.0": { "*": ["ts4.0/*"] }
    }
```

**Comment by DetachHead** [comment]
@angelsvirkov that's intentional yeah. i followed the same pattern the rest of the repo seems to use. that's also how it's shown in the example [here](https://github.com/DefinitelyTyped/DefinitelyTyped#i-want-to-use-features-from-very-new-typescript-versions)

**Pull request merged**
